recipes/flex/all/test_package/conanfile.py

```diff
@@ -1,40 +1,58 @@
-from conans import ConanFile, CMake, tools
-from conans.errors import ConanException
 import os
+import re
+from io import StringIO
+
+from conan import ConanFile
+from conan.tools.build import can_run
+from conan.tools.cmake import CMake, cmake_layout
 
 
 class TestPackageConan(ConanFile):
     settings = "os", "arch", "compiler", "build_type"
-    generators = "cmake"
+    generators = "CMakeToolchain", "CMakeDeps", "VirtualBuildEnv", "VirtualRunEnv"
     test_type = "explicit"
 
     def requirements(self):
         self.requires(self.tested_reference_str)
 
     def build_requirements(self):
-        self.build_requires(self.tested_reference_str)
+        self.tool_requires(self.tested_reference_str)
 
-    def build(self):
-        if not hasattr(self, "settings_build"):
-            # Only test location of flex executable when not cross building
-            flex_bin = tools.which("flex")
-            if not flex_bin.startswith(self.deps_cpp_info["flex"].rootpath):
-                raise ConanException("Wrong flex executable captured")
+    def layout(self):
+        cmake_layout(self)
+
+    def _assert_expected_version(self):
+
+        def tested_reference_version():
+            tokens = re.split('[@#]', self.tested_reference_str)
+            return tokens[0].split("/", 1)[1]
 
-        if not tools.cross_building(self, skip_x64_x86=True) or hasattr(self, "settings_build"):
-            self.run("flex --version", run_environment=not hasattr(self, "settings_build"))
+        output = StringIO()
+        self.run(f"flex --version", output)
+        output_str = str(output.getvalue()) 
+        self.output.info("Installed version: {}".format(output_str))
+        expected_version = tested_reference_version()
+        self.output.info("Expected version: {}".format(expected_version))
+        assert_flex_version = "flex {}".format(expected_version)
+        assert(assert_flex_version in output_str)
 
-            print(os.environ["PATH"])
+    def _create_cpp_file(self):
+        l_file = os.path.join(self.source_folder, "basic_nr.l")
+        output_file = os.path.join(self.build_folder, "basic_nr.cpp")
+        self.run(f"flex --outfile={output_file} --c++ {l_file}")
+        assert os.path.exists(output_file)
+
+    def build(self):
+        if can_run(self):
+            # Let's check flex version installed and create the basic_nr.cpp file
+            self._assert_expected_version()
```

Comment:
I think this could be done outside the `can_run` block - because the `flex` executable that we call should be the one from the `tool_requires` (which will be present in the virtualbuildenv). This would need line 31 above to reference the build environment (`env="conanbuild"` I think)  when calling `self.run()`.

Reply:
Then, I think it does not make much sense to use `can_run()`? Should I remove it?